qiskit/quantum_info/synthesis/xx_decompose/decomposer.py

```diff
@@ -0,0 +1,251 @@
+        weyl_decomposition = TwoQubitWeylDecomposition(unitary)
+        target = [getattr(weyl_decomposition, x) for x in ("a", "b", "c")]
+        if target[-1] < -EPSILON:
+            target = [np.pi / 2 - target[0], target[1], -target[2]]
+
+        best_sequence = self._best_decomposition(target, strengths)["sequence"]
+        return len(best_sequence)
+
+    @staticmethod
+    def _strength_to_infidelity(basis_fidelity, approximate=False):
+        """
+        Converts a dictionary mapping ZX strengths to fidelities to a dictionary mapping ZX
+        strengths to infidelities. Also one of the other formats QISKit uses: extends a single float
+        infidelity over CX, CX/2, and CX/3 if only a single float is supplied.
+        """
+
+        if basis_fidelity is None:
+            basis_fidelity = 1.0
+
+        if isinstance(basis_fidelity, float):
+            if not approximate:
+                slope, offset = 1e-10, 1e-12
+            else:
+                slope, offset = (1 - basis_fidelity) / 2, (1 - basis_fidelity) / 2
+            return {
+                strength: slope * strength / (np.pi / 2) + offset
+                for strength in [np.pi / 2, np.pi / 4, np.pi / 6]
+            }
+        elif isinstance(basis_fidelity, dict):
+            return {
+                strength: (1 - fidelity if approximate else 1e-12 + 1e-10 * strength / (np.pi / 2))
+                for (strength, fidelity) in basis_fidelity.items()
+            }
+
+        raise TypeError("Unknown basis_fidelity payload.")
+
+    def __call__(self, unitary, basis_fidelity=1.0, approximate=True, chatty=False):
+        """
+        Fashions a circuit which (perhaps `approximate`ly) models the special unitary operation `u`,
+        using the circuit templates supplied at initialization.  The routine uses `basis_fidelity`
+        to select the optimal circuit template, including when performing exact synthesis; the
+        contents of `basis_fidelity` is a dictionary mapping interaction strengths (scaled so that
+        CX = RZX(pi/2) corresponds to pi/2) to circuit fidelities.
+        """
+        strength_to_infidelity = self._strength_to_infidelity(
+            basis_fidelity, approximate=approximate
+        )
+
+        # get the associated _positive_ canonical coordinate
+        weyl_decomposition = TwoQubitWeylDecomposition(unitary)
+        target = [getattr(weyl_decomposition, x) for x in ("a", "b", "c")]
+        if target[-1] < -EPSILON:
+            target = [np.pi / 2 - target[0], target[1], -target[2]]
+
+        # scan for the best point
+        best_point, best_sequence = itemgetter("point", "sequence")(
+            self._best_decomposition(target, strength_to_infidelity)
+        )
+        # build the circuit building this canonical gate
+        embodiments = {
+            k: self.embodiments.get(k, self._default_embodiment(k))
+            for k, v in strength_to_infidelity.items()
+        }
```

Comment:
somewhere can you check that the embodiment that users build are actually correct? this is to prevent users making mistakes without knowing. No need to check for _default_embodiment since that is in the tests.

Reply:
Added a check at initialization time.